File: packages/server/src/crypto.ts
/**
 * Cryptographic utilities for webhook signatures and callback tokens.
 */

import {
  createHash,
  createHmac,
  createPublicKey,
  generateKeyPairSync,
  randomBytes,
  sign,
  timingSafeEqual,
  verify as verifySignature,
} from "node:crypto"
import type { JsonWebKey as NodeJsonWebKey } from "node:crypto"

export interface WebhookPublicJwk {
  kty: `OKP`
  crv: `Ed25519`
  x: string
  kid: string
  use: `sig`
  alg: `EdDSA`
}

export interface WebhookJwks {
  keys: Array<WebhookPublicJwk>
}

/**
 * Generate a webhook secret for a subscription.
 */
export function generateWebhookSecret(): string {
  return `whsec_${randomBytes(32).toString(`hex`)}`
}

/**
 * Generate a unique wake ID.
 */
export function generateWakeId(): string {
  return `w_${randomBytes(12).toString(`hex`)}`
}

const WEBHOOK_KEYPAIR = generateKeyPairSync(`ed25519`)
const WEBHOOK_PUBLIC_JWK = buildWebhookPublicJwk()

function buildWebhookPublicJwk(): WebhookPublicJwk {
  const exported = WEBHOOK_KEYPAIR.publicKey.export({ format: `jwk` }) as {
    kty?: string
    crv?: string
    x?: string
  }

  if (exported.kty !== `OKP` || exported.crv !== `Ed25519` || !exported.x) {
    throw new Error(`Failed to export Ed25519 webhook signing key`)
  }

  const thumbprintInput = JSON.stringify({
    crv: exported.crv,
    kty: exported.kty,
    x: exported.x,
  })
  const kid = `ds_${createHash(`sha256`)
    .update(thumbprintInput)
    .digest(`base64url`)}`

  return {
    kty: `OKP`,
    crv: `Ed25519`,
    x: exported.x,
    kid,
    use: `sig`,
    alg: `EdDSA`,
  }
}

export function getWebhookSigningKeyId(): string {
  return WEBHOOK_PUBLIC_JWK.kid
}

export function getWebhookJwks(): WebhookJwks {
  return {
    keys: [{ ...WEBHOOK_PUBLIC_JWK }],
  }
}

/**
 * Sign a webhook payload for the Webhook-Signature header.
 *
 * Without a secret, signs with the upstream Ed25519/JWKS scheme.
 * With a secret, signs with the PR #343 HMAC scheme used by the
 * layered webhook conformance tests.
 */
export function signWebhookPayload(body: string): string
export function signWebhookPayload(body: string, secret: string): string
export function signWebhookPayload(body: string, secret?: string): string {
  const timestamp = Math.floor(Date.now() / 1000)
  const payload = `${timestamp}.${body}`

  if (secret) {
    const signature = createHmac(`sha256`, secret).update(payload).digest(`hex`)
    return `t=${timestamp},sha256=${signature}`
  }

  const signature = sign(
    null,
    Buffer.from(payload),
    WEBHOOK_KEYPAIR.privateKey
  ).toString(`base64url`)
  return `t=${timestamp},kid=${WEBHOOK_PUBLIC_JWK.kid},ed25519=${signature}`
}

/**
 * Verify a webhook signature.
 *
 * Passing a string verifies the PR #343 HMAC scheme. Passing a JWKS, or
 * omitting the third argument, verifies the upstream Ed25519/JWKS scheme.
 */
export function verifyWebhookSignature(
  body: string,
  signatureHeader: string,
  jwks?: WebhookJwks,
  toleranceSeconds?: number
): boolean
export function verifyWebhookSignature(
  body: string,
  signatureHeader: string,
  secret: string,
  toleranceSeconds?: number
): boolean
export function verifyWebhookSignature(
  body: string,
  signatureHeader: string,
  verifier: WebhookJwks | string = getWebhookJwks(),
  toleranceSeconds = 300
): boolean {
  if (typeof verifier === `string`) {
    const match = signatureHeader.match(/t=(\d+),sha256=([a-f0-9]+)/)
    if (!match) return false

    const [, timestamp, signature] = match
    const ts = parseInt(timestamp!, 10)
    const now = Math.floor(Date.now() / 1000)
    if (Math.abs(now - ts) > toleranceSeconds) return false

    const payload = `${timestamp}.${body}`
    const expected = createHmac(`sha256`, verifier)
      .update(payload)
      .digest(`hex`)

    try {
      return timingSafeEqual(Buffer.from(signature!), Buffer.from(expected))
    } catch {
      return false
    }
  }

  const match = signatureHeader.match(
    /^t=(\d+),kid=([^,]+),ed25519=([A-Za-z0-9_-]+)$/
  )
  if (!match) return false

  const [, timestamp, kid, signature] = match
  const ts = parseInt(timestamp!, 10)
  const now = Math.floor(Date.now() / 1000)
  if (Math.abs(now - ts) > toleranceSeconds) return false

  const jwk = verifier.keys.find((key) => key.kid === kid)
  if (!jwk) return false

  try {
    const publicKey = createPublicKey({
      key: jwk as unknown as NodeJsonWebKey,
      format: `jwk`,
    })
    return verifySignature(
      null,
      Buffer.from(`${timestamp}.${body}`),
      publicKey,
      Buffer.from(signature!, `base64url`)
    )
  } catch {
    return false
  }
}

// Token signing key — generated per server instance
const TOKEN_KEY = randomBytes(32)

/**
 * Generate a signed callback token.
 * Token format: base64url(json_payload).base64url(hmac_signature)
 * Payload: { consumer_id, epoch, exp }
 */
export function generateCallbackToken(
  consumerId: string,
  epoch: number
): string {
  const payload = {
    sub: consumerId,
    epoch,
    exp: Math.floor(Date.now() / 1000) + 3600, // 1 hour TTL
    jti: randomBytes(8).toString(`hex`),
  }
  const payloadStr = Buffer.from(JSON.stringify(payload)).toString(`base64url`)
  const sig = createHmac(`sha256`, TOKEN_KEY)
    .update(payloadStr)
    .digest(`base64url`)
  return `${payloadStr}.${sig}`
}

/** Seconds before expiry at which a token should be refreshed. */
const TOKEN_REFRESH_THRESHOLD = 300 // 5 minutes

/**
 * Validate a callback token. Returns the decoded payload or null.
 * On success, includes `exp` (unix seconds) so callers can decide
 * whether the token needs refreshing.
 */
export function validateCallbackToken(
  token: string,
  consumerId: string
):
  | { valid: true; exp: number; epoch: number }
  | { valid: false; code: `TOKEN_INVALID` | `TOKEN_EXPIRED` } {
  const parts = token.split(`.`)
  if (parts.length !== 2) {
    return { valid: false, code: `TOKEN_INVALID` }
  }

  const [payloadStr, sig] = parts

  const expectedSig = createHmac(`sha256`, TOKEN_KEY)
    .update(payloadStr!)
    .digest(`base64url`)

  try {
    if (!timingSafeEqual(Buffer.from(sig!), Buffer.from(expectedSig))) {
      return { valid: false, code: `TOKEN_INVALID` }
    }
  } catch {
    return { valid: false, code: `TOKEN_INVALID` }
  }

  let payload: { sub: string; epoch: number; exp: number }
  try {
    payload = JSON.parse(Buffer.from(payloadStr!, `base64url`).toString())
  } catch {
    return { valid: false, code: `TOKEN_INVALID` }
  }

  if (payload.sub !== consumerId) {
    return { valid: false, code: `TOKEN_INVALID` }
  }

  const now = Math.floor(Date.now() / 1000)
  if (now > payload.exp) {
    return { valid: false, code: `TOKEN_EXPIRED` }
  }

  return { valid: true, exp: payload.exp, epoch: payload.epoch }
}

/**
 * Check whether a token is close enough to expiry that it should be refreshed.
 */
export function tokenNeedsRefresh(exp: number): boolean {
  const now = Math.floor(Date.now() / 1000)
  return exp - now <= TOKEN_REFRESH_THRESHOLD
}

File: packages/caddy-plugin/webhook/crypto.go
package webhook

import (
	"crypto/ed25519"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// tokenKey is generated once per process for signing callback tokens.
var tokenKey []byte
var webhookPublicKey ed25519.PublicKey
var webhookPrivateKey ed25519.PrivateKey
var webhookPublicJWK WebhookPublicJWK

func init() {
	tokenKey = make([]byte, 32)
	if _, err := rand.Read(tokenKey); err != nil {
		panic(fmt.Sprintf("failed to generate token key: %v", err))
	}

	var err error
	webhookPublicKey, webhookPrivateKey, err = ed25519.GenerateKey(rand.Reader)
	if err != nil {
		panic(fmt.Sprintf("failed to generate webhook signing key: %v", err))
	}
	webhookPublicJWK = buildWebhookPublicJWK(webhookPublicKey)
}

const tokenRefreshThreshold = 300 // 5 minutes in seconds

// WebhookPublicJWK is the public Ed25519 webhook signing key in JWK form.
type WebhookPublicJWK struct {
	Kty string `json:"kty"`
	Crv string `json:"crv"`
	Kid string `json:"kid"`
	Use string `json:"use"`
	Alg string `json:"alg"`
	X   string `json:"x"`
}

// WebhookJWKS is the JSON Web Key Set served to webhook receivers.
type WebhookJWKS struct {
	Keys []WebhookPublicJWK `json:"keys"`
}

func buildWebhookPublicJWK(publicKey ed25519.PublicKey) WebhookPublicJWK {
	x := base64.RawURLEncoding.EncodeToString(publicKey)
	thumbprintInput := fmt.Sprintf(`{"crv":"Ed25519","kty":"OKP","x":"%s"}`, x)
	thumbprint := sha256.Sum256([]byte(thumbprintInput))

	return WebhookPublicJWK{
		Kty: "OKP",
		Crv: "Ed25519",
		Kid: "ds_" + base64.RawURLEncoding.EncodeToString(thumbprint[:]),
		Use: "sig",
		Alg: "EdDSA",
		X:   x,
	}
}

// GetWebhookSigningKeyID returns the active webhook signing key ID.
func GetWebhookSigningKeyID() string {
	return webhookPublicJWK.Kid
}

// GetWebhookJWKS returns the active webhook signing public key set.
func GetWebhookJWKS() WebhookJWKS {
	return WebhookJWKS{
		Keys: []WebhookPublicJWK{webhookPublicJWK},
	}
}

// GenerateWakeID creates a unique wake ID prefixed with "w_".
func GenerateWakeID() string {
	b := make([]byte, 12)
	rand.Read(b)
	return "w_" + hex.EncodeToString(b)
}

// SignWebhookPayload signs a webhook body with the server webhook signing key.
// Returns "t=<unix_ts>,kid=<key_id>,ed25519=<base64url_sig>".
func SignWebhookPayload(body string) string {
	timestamp := time.Now().Unix()
	payload := fmt.Sprintf("%d.%s", timestamp, body)
	sig := ed25519.Sign(webhookPrivateKey, []byte(payload))
	return fmt.Sprintf(
		"t=%d,kid=%s,ed25519=%s",
		timestamp,
		webhookPublicJWK.Kid,
		base64.RawURLEncoding.EncodeToString(sig),
	)
}

// tokenPayload is the internal structure of a callback token.
type tokenPayload struct {
	Sub   string `json:"sub"`
	Epoch int    `json:"epoch"`
	Exp   int64  `json:"exp"`
	Jti   string `json:"jti"`
}

// GenerateCallbackToken creates a signed callback token for a consumer.
func GenerateCallbackToken(consumerID string, epoch int) string {
	jti := make([]byte, 8)
	rand.Read(jti)

	payload := tokenPayload{
		Sub:   consumerID,
		Epoch: epoch,
		Exp:   time.Now().Unix() + 3600, // 1 hour TTL
		Jti:   hex.EncodeToString(jti),
	}

	payloadJSON, _ := json.Marshal(payload)
	payloadStr := base64.RawURLEncoding.EncodeToString(payloadJSON)

	mac := hmac.New(sha256.New, tokenKey)
	mac.Write([]byte(payloadStr))
	sig := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))

	return payloadStr + "." + sig
}

// TokenValidation is the result of validating a callback token.
type TokenValidation struct {
	Valid bool
	Exp   int64
	Code  string // "TOKEN_INVALID" or "TOKEN_EXPIRED" when !Valid
}

// ValidateCallbackToken verifies a callback token and returns the validation result.
func ValidateCallbackToken(token, consumerID string) TokenValidation {
	parts := strings.SplitN(token, ".", 2)
	if len(parts) != 2 {
		return TokenValidation{Valid: false, Code: ErrCodeTokenInvalid}
	}

	payloadStr, sig := parts[0], parts[1]

	// Verify HMAC
	mac := hmac.New(sha256.New, tokenKey)
	mac.Write([]byte(payloadStr))
	expectedSig := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))

	if !hmac.Equal([]byte(sig), []byte(expectedSig)) {
		return TokenValidation{Valid: false, Code: ErrCodeTokenInvalid}
	}

	// Decode payload
	payloadJSON, err := base64.RawURLEncoding.DecodeString(payloadStr)
	if err != nil {
		return TokenValidation{Valid: false, Code: ErrCodeTokenInvalid}
	}

	var payload tokenPayload
	if err := json.Unmarshal(payloadJSON, &payload); err != nil {
		return TokenValidation{Valid: false, Code: ErrCodeTokenInvalid}
	}

	if payload.Sub != consumerID {
		return TokenValidation{Valid: false, Code: ErrCodeTokenInvalid}
	}

	now := time.Now().Unix()
	if now > payload.Exp {
		return TokenValidation{Valid: false, Code: ErrCodeTokenExpired}
	}

	return TokenValidation{Valid: true, Exp: payload.Exp}
}

// TokenNeedsRefresh returns true if the token is within 5 minutes of expiry.
func TokenNeedsRefresh(exp int64) bool {
	now := time.Now().Unix()
	return exp-now <= tokenRefreshThreshold
}
